### DISCUSSION TITLE
rpc

### DISCUSSION THREAD

**Discussion by manson** [pocketbase/pocketbase#86] [Q&A] [answered]
I see a good potential for an embedded usage in my use cases. But as always happens there are some small things that seems to make it even more perfect if to be implemented.   
1) To me the small and mid-size projects are better to have document-based storage (bolt, badger, mongo) or some other similar kv or doc storages to skip this annoying migrations. To tell the truth it would be great to abstract db usage the user can do what he wants and leave author alone with this topic :-) But ok, for small/mid projects sqlite is ok. 
2) Its great to have existing http hooks to implement my own server services but it is lack of rpc (json, http, proto or whatever). It is a great point to make this app a bit more distributed (interconnected the same nodes). It is a possibility but user can do whatever he wants. As for me, I plan a small project with a server-side scripting with golang language. I suppose to create and use something like stored procedures in big sql servers but not limited only with data. And it is best to use RPC for me (http-rpc for browser clients) to remotely call this functions (and update/delete/add). Other than http rpc is pretty simple to implement as it will reside on different port but with http its better to implement into the core. But maybe there are rpc middlewares for echo you use? It would be great...

**Comment by ganigeorgiev** [comment, marked as answer]
1. bolt, badger, pebble, etc. are very basic key-value stores and will require a lot of effort to implement the query mechanism and syntax of SQL (which developers already know). I actually experimented at beginning with bold, badger, badgerhold, genji, lungo and other databases and in all of my local tests SQLite had superb performance and extremely low memory consumption. Support for other SQL databases could be implemented in the future (see #14), but it's a low priority for now. NoSQL databases are not planned because we rely heavily on the SQL syntax (even for the collection access filters).

3. PocketBase cannot be distributed out of the box. It is just a go library/framework/toolkit and nothing is stopping you to start a rpc/grpc server in a new goroutine and integrate it with the PocketBase HTTP API or use the internal database. Lower-level network features like support to bind directly to unix socket could be added, but I'm not sure how many people will use it, so for now is not in the roadmap.

**Reply by manson** [reply]
1) Not sure sqlite outperforms doc-based kv storages very much especially if some of them were developed for speed and because of their simplicity. SQL queries - yes, this is the  advantage but we're talking about the small and mid-sized applications, so in this case this is not a big deal. But migrating is something very lame, uncomfortable and point of different exceptions and headaches. I've been working a lot with a big systems (ms sql server) hundreds and thousands tables and yes - without sql it would be impossible to process those data, but... When I left big organization I never had a necessity to use regular relational storage. If transaction is not the first citizen of application, a simple and fast document database (or key-value) always was enough and very simple to maintain. But I understand you. But if there was some simple interface to this in core even without implementation it would be great.

2) Yea. I found an echo rpc middleware, so its great. Will be thinking further how I could use this in my cases...

Thanks! Go further! Will watch your progress! 👍

**Reply by pnmcosta** [reply]
@manson could you share an update on rpc middleware? And which you ended up going with? Thanks.
